### PR TITLE
FeedbackRubricQuestionDetails: Refactor calculateRubricStats #6558

### LINKED ISSUE #6558: FeedbackRubricQuestionDetails: Refactor calculateRubricStats
Current implementation of calculateRubricStats calls calculateResponseFrequency within the method, which is already always called before calling calculateRubricStats. The two methods should be refactored to not call calculateResponseFrequency twice.

### PR DESCRIPTION
**Pull request by leeyimin**
Fixes #6558

**Outline of Solution**
The two methods are extracted into a nested static class. 
Calculation for both arrays are done in the constructor of the nested static class.

### CODE REVIEW

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 1]
```diff
@@ -964,4 +895,83 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
         return rubricSubQuestions;
     }
 
+    /**
+     * Calculates the response frequency, percentage frequency and average value
+     * for each choice and each sub-question given the responses and question details
+     *
+     * For responseFrequency, Last element in each row stores the total number of responses for the sub-question.
```
Comment (samsontmr) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: Last -> last

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 2]
```diff
@@ -964,4 +895,83 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
...
+            for (FeedbackResponseAttributes response : responses) {
+                FeedbackRubricResponseDetails frd = (FeedbackRubricResponseDetails) response.getResponseDetails();
+                for (int i = 0; i < numOfRubricSubQuestions; i++) {
+                    int chosenChoice = frd.getAnswer(i);
+                    if (chosenChoice != -1) {
+                        responseFrequency[i][chosenChoice] += 1;
+                        responseFrequency[i][responseTotalIndex] += 1;
+                    }
+                }
+            }
```
Comment (samsontmr) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: Consider extracting this for loop into a separate method.

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 3]
```diff
@@ -964,4 +895,83 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
...
+                }
+                // Calculate the average for each sub-question
+                if (questionDetails.hasAssignedWeights) {
+                    for (int j = 0; j < numOfRubricChoices; j++) {
+                        float choiceWeight =
+                                (float) (questionDetails.rubricWeights.get(j) * percentageFrequencyAndAverage[i][j]);
+                        percentageFrequencyAndAverage[i][numOfRubricChoices] += choiceWeight;
+                    }
+                }
+            }
```
Comment (samsontmr) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: Consider extracting parts of this for loop into separate methods.

Comment (leeyimin) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: Chose not to further extract parts of the loop out as the code is relatively simply and the second inner loop depends on the first. There are not much benefits in further extracting (If second inner loop is extracted out as it is, readers have jump to another method to fully understand the code. If the second inner loop is extracted out with its own outer loop, the code would require the percentage frequencies to be computed first, introducing additional dependency.)

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 4]
```diff
@@ -941,7 +953,15 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
...
             }
+        }
 
+        /**
+         * Calculates the percentage frequencies for each choice and average value for each sub-question
+         * Requires responseFrequency to be computed
+         */
+        void calculatePercentageFrequencyAndAverage() {
+            percentageFrequencyAndAverage =
+                    new float[numOfRubricSubQuestions][numOfRubricChoices + 1];
```
Comment (samsontmr) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: Make them on the same line if possible?

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 5]
```diff
@@ -919,17 +919,29 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
...
+            this.numOfRubricChoices = questionDetails.numOfRubricChoices;
+            this.responseTotalIndex = numOfRubricChoices;
 
+            calculateResponseFrequency();
+            calculatePercentageFrequencyAndAverage();
+        }
+
+        void calculateResponseFrequency() {
             responseFrequency =
                     new int[numOfRubricSubQuestions][numOfRubricChoices + 1];
```
Comment (samsontmr) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: this too

**Comment by leeyimin** [comment]
@wkurniawan07 Ready for review

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 6]
```diff
@@ -964,4 +895,101 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
...
+     * -> is the total number of the responses for the given sub-question
+     *
+     * For percentageFrequencyAndAverage, values are set to 0 if there are no responses to that sub-question.
+     * Average value is set to 0 if there are no assigned weights.
+     * e.g.
+     * percentageFrequencyAndAverageValue[subQuestionIndex][choiceIndex]
+     * -> is the percentage choiceIndex is chosen for subQuestionIndex
+     * percentageFrequencyAndAverageValue[subQuestionIndex][numOfRubricChoices]
+     * -> is the average weight of the responses for the given sub-question
+     */
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: This is too detailed for a header comment. To explain the class, something like "Statistics for a rubric question, consisting of ...".
The explanations are more apt for the variables:
```java
/**
 * Last element in each row ...
 * ... the given sub-question.
 */
int[][] responseFrequency;

/**
 * Values are set to 0 ...
 * ... the given sub-question.
 */
float[][] percentageFrequencyAndAverage;
```

**Comment by leeyimin** [comment]
@wkurniawan07 Changes made

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 7]
```diff
@@ -964,4 +895,103 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
...
+                        responseFrequency[i][chosenChoice] += 1;
+                        responseFrequency[i][responseTotalIndex] += 1;
+                    }
+                }
+            }
+        }
+
+        /**
+         * Calculates the percentage frequencies for each choice and average value for each sub-question
+         * Requires responseFrequency to be computed
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: Is it possible to verify this precondition using an `Assumption.*`?

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 8]
```diff
@@ -964,4 +895,103 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
...
+                int totalForSubQuestion = responseFrequency[i][responseTotalIndex];
+                //continue to next row if no response for this sub-question
+                if (totalForSubQuestion == 0) {
+                    continue;
+                }
+                // Divide responsesFrequency by totalForSubQuestion to get percentage
+                for (int j = 0; j < numOfRubricChoices; j++) {
+                    percentageFrequencyAndAverage[i][j] = (float) responseFrequency[i][j] / totalForSubQuestion;
+                }
+                // Calculate the average for each sub-question
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: To be consistent, all short comments like this should start with a lower case letter?

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 9]
```diff
@@ -964,4 +895,103 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
         return rubricSubQuestions;
     }
 
+    /**
+     * Class to calculate the statistics of responses for a rubric question
+     */
+    private static class RubricStatistics {
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: Did you consider the choices static nested class vs non-static inner class?

Comment (leeyimin) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: I did not consider taking out the questionDetails object at first so I thought static nested was a better choice as it shows that the RubricStatistics class did not depend on other variables of the FeedbackRubricQuestionDetails object calling it. 
I just realised that it is possible to construct a inner class object from questionDetails object instead but the syntax is not very nice and potentially confusing.
```
RubricStatistics statistics = fqd.new RubricStatistics(responsesForStatistics);
```

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java (hunk 7 above)
Comment (leeyimin) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: Added a not null assertion on responseFrequency to ensure it is at least initialised (which is currently in calculateResponseFrequency).

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 10]
```diff
@@ -964,4 +896,106 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
...
+            this.numOfRubricChoices = questionDetails.numOfRubricChoices;
+            this.responseTotalIndex = numOfRubricChoices;
+
+            calculateResponseFrequency();
+            calculatePercentageFrequencyAndAverage();
+        }
+
+        void calculateResponseFrequency() {
+            responseFrequency = new int[numOfRubricSubQuestions][numOfRubricChoices + 1];
+            // Count frequencies
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: A few other places where one line comments start with upper case letter

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java (hunk 7 above)
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: Rephrase as `Precondition: responseFrequency has been calculated`?

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 11]
```diff
@@ -964,4 +896,106 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
...
+         * e.g.
+         * responseFrequency[subQuestionIndex][choiceIndex]
+         * -> is the number of times choiceIndex is chosen for subQuestionIndex
+         * responseFrequency[subQuestionIndex][numOfRubricChoices]
+         * -> is the total number of the responses for the given sub-question
+         */
+        int[][] responseFrequency;
+
+        /**
+         * Values are set to 0 if there are no responses to that sub-question.
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: Perhaps start with a overview sentence (i.e. what is this variable about?) before explaining finer details? Applies to the variable above as well.

**Comment by leeyimin** [comment]
@damithc Ready for review!

File: src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java [hunk 12]
```diff
@@ -964,4 +896,110 @@ public void setNumOfRubricSubQuestions(int numOfRubricSubQuestions) {
...
     }
 
+    /**
+     * Class to calculate the statistics of responses for a rubric question
+     */
+    private static class RubricStatistics {
+
+        /**
+         * Stores the frequency of being selected for each choice of each sub-question
+         * and the total number of responses for each sub-question
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackRubricQuestionDetails.java]: nit: full stop missing. Applies to other comments too.

**Pull request merged**